### PR TITLE
FTP uploader plugin

### PR DESCRIPTION
**Pull request by moix**
This simple plugin add the capability to upload output to a ftp

### CODE REVIEW

**Comment by soulgalore** [comment]
Hey @moix thanks for the PR! 

A standard ftp-plugin is a really good idea!

It's the linting that fails: https://travis-ci.org/sitespeedio/sitespeed.io/builds/297368274?utm_source=github_status&utm_medium=notification

Can you fix that and I'll have a look? `npm run lint` and best case `npm run lint:fix`.

Best
Peter

**Comment by moix** [comment]
hey @soulgalore thanks! sorry I'm really a newbie here, can you point me how to setup npm lint? eslint fails due missing conf file. I tried `eslint --init` but dunno what options should put
Best /moix

**Comment by soulgalore** [comment]
Hey @moix hmm I wonder if something wrong in our setup, ca share the exact error you get (ping @tobli who has better knowledge of how it should be done).

@moix I think this is a really good PR, I'll just add a couple of things that need to be fixed to follow best practices ok?

File: lib/plugins/ftp/index.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+const merge = require('lodash.merge');
+const throwIfMissing = require('../../support/util').throwIfMissing;
+const fs = require('fs-extra');
+const defaultConfig = {
+  host: 'localhost',
+  port: 21,
+  user: 'user',
+  password: 'password'
+};
+function createFTPClient(ftpOptions) {
```
Comment (soulgalore) [inline lib/plugins/ftp/index.js]: Is this function used anywhere, i cannot find it? The ftpOptions is already defined in the function body as const, you can just rename it.

File: lib/plugins/ftp/index.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    this.storageManager = context.storageManager;
+  },
+
+  processMessage(message, queue) {
+    const make = this.make;
+    if (message.type === 'html.finished') {
+      const ftpOptions = merge({}, defaultConfig, this.ftpOptions);
+      const baseDir = this.storageManager.getBaseDir();
+      const storagePathPrefix = this.storageManager.getStoragePrefix();
+      var client = new PromiseFtp();
```
Comment (soulgalore) [inline lib/plugins/ftp/index.js]: All var can be const, so just change all vars to be const :)

File: lib/plugins/ftp/index.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+      return new Promise((resolve, reject) => {
+        log.info(`FTP uploading output (${baseDir}) to FTP ${ftpOptions.user}@${ftpOptions.host}, this can take a while ...`);
+        var dirlist = [];
+        var filelist = [];
+
+        function walkSync(dir) {
+          var path = path || require('path');
```
Comment (soulgalore) [inline lib/plugins/ftp/index.js]: We usually do all requires at the top of the file to make it easy to see what's used, so if you can move it, thanks!

File: lib/plugins/ftp/index.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+          var fs = fs || require('fs'),
```
Comment (soulgalore) [inline lib/plugins/ftp/index.js]: We have fs-extra already we can use that one right?

File: lib/plugins/ftp/index.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+const throwIfMissing = require('../../support/util').throwIfMissing;
+const fs = require('fs-extra');
+const defaultConfig = {
+  host: 'localhost',
+  port: 21,
+  user: 'user',
+  password: 'password'
+};
+function createFTPClient(ftpOptions) {
+  const ftpOptions = merge({}, defaultConfig, options.ftp);
```
Comment (soulgalore) [inline lib/plugins/ftp/index.js]: ftpOptions is already defined as the parameter.

**Comment by moix** [comment]
Hi @soulgalore thanks for your comments about good practices, I have added them.
@tobli any other suggestion about what can be failing during build? there was a push with some lint failures [build 297866565](https://travis-ci.org/sitespeedio/sitespeed.io/builds/297866565?utm_source=github_status&utm_medium=notification), I fixed them but now is failing with initial failure we had yesterday [build 297872371](https://travis-ci.org/sitespeedio/sitespeed.io/builds/297872371?utm_source=github_status&utm_medium=notification)

**Comment by tobli** [comment]
Prettier is run as a eslint plugin. You first need to run npm install to install the versions of eslint an prettier that's defined in package.json. Try this:
```
# install dependencies (eslint, prettier etc.)
npm install
# check (and fix) code style
npm run lint:fix
```

You'll see one eslint error for an unused 'resolve' function, i.e. there's a promise that's never resolved. Hence the message processing will hang in the queue.

**Pull request closed without merging**